src/org/exist/source/SourceFactory.java

```diff
@@ -58,19 +64,26 @@
      * @throws MalformedURLException
      * @throws IOException
      */
-    public static final Source getSource(DBBroker broker, String contextPath, String location, boolean checkXQEncoding) throws MalformedURLException, IOException, PermissionDeniedException
-    {
+    public static final Source getSource(DBBroker broker, String contextPath, String location, boolean checkXQEncoding) throws IOException, PermissionDeniedException {
         Source source = null;
-        
+
+        /* resource: */
+        if(location.startsWith(ClassLoaderSource.PROTOCOL)) {
+            source = new ClassLoaderSource(location);
+        } else if(contextPath != null && contextPath.startsWith(ClassLoaderSource.PROTOCOL)) {
+            // Pretend it is a file on the local system so we can resolve it easily with URL() class.
+            final String conPathNoProtocol = contextPath.replace(ClassLoaderSource.PROTOCOL, "file://");
+            String resolvedURL = new URL(new URL(conPathNoProtocol), location).toString();
+            resolvedURL = resolvedURL.replaceFirst("file://", ClassLoaderSource.PROTOCOL);
+            source = new ClassLoaderSource(resolvedURL);
+        }
         /* file:// or location without scheme is assumed to be a file */
-        if(location.startsWith("file:") || location.indexOf(':') == Constants.STRING_NOT_FOUND)
+        else if(location.startsWith("file:") || location.indexOf(':') == Constants.STRING_NOT_FOUND)
```

Comment:
Constants.STRING_NOT_FOUND ... can be replaced with a Java6+ operation on the string object?